drivers/i2c/i2c_ll_stm32_v1.c

```diff
@@ -653,13 +657,15 @@ s32_t stm32_i2c_msg_write(struct device *dev, struct i2c_msg *msg,
 	struct i2c_stm32_data *data = DEV_DATA(dev);
 	I2C_TypeDef *i2c = cfg->i2c;
 	u32_t len = msg->len;
-	u16_t timeout;
+	const s32_t timeout_us = (cfg->timeout == K_FOREVER) ? K_FOREVER :
```

Comment:
IIUC we're extending the blocking behavior as default case for poling mode.
I have 2 concerns with this change:
- default behavior should be to error on timeout, and not to block.
- we're modifying behavior for existing users of the API.

I'm ok with adding the possibility to set a timeout value by dt, but we should not make "blocking" a default behavior.

Reply:
@erwango "Blocking" is the default behavior for v2 driver and v1 driver interrupt mode. Only v1 driver polling mode has a timeout. In order not to modify behavior for existing users of the API in 3 out of 4 cases I keep the "blocking" default behavior.

`timeout-ms` property is optional and defaults to `-1 (K_FOREVER)` now, in the future we could easily change it to something else editing the yaml files.

Reply:
> @erwango "Blocking" is the default behavior for v2 driver and v1 driver interrupt mode. Only v1 driver polling mode has a timeout. In order not to modify behavior for existing users of the API in 3 out of 4 cases I keep the "blocking" default behavior.

Ok, fair enough but I was actually considering that "blocking" was an faulty behavior, and that #21293 should rather be a bug, since blocking on error is not a proper behavior.

> timeout-ms property is optional and defaults to -1 (K_FOREVER) now, in the future we could easily change it to something else editing the yaml files.

I think should be the direction indeed.

Reply:
> Ok, fair enough but I was actually considering that "blocking" was an faulty behavior, and that #21293 should rather be a bug, since blocking on error is not a proper behavior.

I haven't checked the other I2C drivers but the I2C API doesn't give the option of timeout in a transaction.

Reply:
Indeed. Unfortunately we miss a I2C API owner to discuss this change further.

@carlescufi, maybe you can help since you're used to API discussions:
Are there general recommendation on API behavior in case of errors?
On I2C at least this is not very well documented and on STM32 we're blocking and never return error. On some other I2C drivers it seems to be a possible behavior as well, while it doesn't seem an adequate behavior.

Reply:
@erwango no, there is no generic handling of errors as far as I know, beyond the obvious of returning an error code. But in this particular case maybe @anangl or @pabigot can provide feedback on the best course of action.

Reply:
I think an issue should be opened to add timeout support to I2C generically, and consider how it could be done for other asynchronous operations that may fail to complete.  I suspect the solution will require an ability to abort an operation, and to reset the device to a known-good state.  For I2C that would include the manual "clear-the-bus" protocol that has an issue somewhere.

Reply:
> beyond the obvious of returning an error code

@carlescufi, @pabigot , since current behavior is blocking we're obvioulsy not returning an error. OTOH, if there is no generic timeout support, would it mean this is acceptable for now? (I'd say "no", but I'm open to keep blocking behavior if that's the consensus)

Reply:
My opinion:  #21538 is pending for asynchronous APIs, I intend to work this in January after GPIO is settled.

At this time I2C is synchronous only.  If ST wants to provide a timeout facility that doesn't change the API, except to return an error indicator if a timeout occurs, I won't block it, but I would prefer to solve the problem generically.

The generic case requires an API change to configure timeout at runtime, possibly per-transaction.  It may be better to keep things as they are until we figure out what that API should look like, so we don't introduce an ST-specific behavior that has to be changed to become consistent with the generic behavior we haven't defined yet.

Reply:
> Since current behavior is blocking we're obvioulsy not returning an error.

Just to make sure that we are on the same page. STM32 I2C driver is blocking until the data transfer is completed or an error occurs (e.g. ARLO, BERR, OVR). The driver blocks forever in case the target device stops operating correctly (e.g. holds SDA line low). In this case the proposed changes allow the application to take some actions (e.g. reset the target device).